### ISSUE TITLE
Add a run exports?

### ISSUE THREAD

**Issue by hmaarrfk** [conda-forge/numpy-feedstock#140]
Should numpy have a run_exports?

It would help with numpy pinnings right?

```
max(min_required_versed, conda_forge_numpy_pinning)
```

this would be so that users could

```
- numpy {{ numpy_version_at_least('1.11') }}
```
but could return 1.15 if numpy 1.11 through 1.14 isn't built.

**Comment by rgommers** [comment]
I've read https://docs.conda.io/projects/conda-build/en/latest/source/resources/define-metadata.html?highlight=run_exports but that's not enough to understand your question here.

> `numpy {{ numpy_version_at_least('1.11') }}`

I'm missing some context I think. That's not quite pinning, what's the difference with `numpy>='1.11'`?

**Comment by hmaarrfk** [comment]
the run_exports question was more about the applicability of `run_exports` to the numpy feedstock. 

For C feedstocks, it kinda makes sense, but maybe it doesn't make sense for numpy recipe users.

with run exports, a recipe would say
```
- requirements
  - host
    - numpy 1.11
```

and thats it, the `pin_compatible` would be inferred by anybody building with the updated recipe.

during build, `numpy>=1.11` will find 1.16, meaning that it won't be compatible with anything built with numpy 1.14. We are mostly struggling with the aarch migration. I would prefer it if we didn't have to change a bunch of recipes for people saying
```
- numpy 1.13
```
to 
```
- numpy 1.13  # [not (aarch64 or ppc64le)]
- numpy 1.15  # [aarch64 or ppcle]
```
if we are for example unable to build numpy 1.13. That would increase the burden on downstream maintainers, for the aarch experiment.

**Comment by rgommers** [comment]
So right now, a package that builds against NumPy (i.e. using the C API), needs to do:
```
- requirements
  - host
    - numpy
  - run
    - {{ pin_compatible('numpy') }}
```
There is no version specifier there for numpy, because the package simply needs to get built for every numpy version that is available. (I verified this for SciPy and PyWavelets; my understanding is that it's generally the case because the whole stack gets rebuilt for every NumPy version).

For a package that just *uses* numpy at runtime, it is
```
- requirements:
  - run:
    - numpy >= 1.13
````

Then third case is end users - they can have `numpy == 1.13` or similar in an `environment.yml` file. 

I think you're talking about cases for building against NumPy where one wants to say `numpy == x.y.z` - when do you do that? Can you point me to an example?

**Comment by hmaarrfk** [comment]
I think it **was** the case that it would get built once per version of numpy. But recently numpy has become forward compatible. So building with 1.11 would be valid for 1.11, 1.12, 1.13, 1.14, 1.15 and 1.16

This is what scikit-image does (simple recipe that uses Cython + numpy).

https://github.com/conda-forge/scikit-image-feedstock

The issue is that scikit-image pins to numpy 1.11, and that might not be able to build for PPC64LE.

wiith run exports, we could remove Line 45 off the recipe
https://github.com/conda-forge/scikit-image-feedstock/blob/master/recipe/meta.yaml#L45

I think most of these run_exports errors are fixed now.

**Comment by rgommers** [comment]
> I think it was the case that it would get built once per version of numpy. But recently numpy has become forward compatible. 

That has been the case before conda was invented; the last serious forward compatibility break was in numpy 1.4.1 in 2010. There have been minor issues a few times, but those were typically quickly corrected.

**Comment by rgommers** [comment]
Okay, the scikit-image example helps, thanks.

It's not quite clear to me that this is desirable though. For example, Travis O has been arguing for being more flexible about breaking the NumPy ABI exactly because conda solves the issue of having to be forward-compatible. Rebuilds for every numpy version is arguably a feature not a limitation. It does cause extra builds of course, but it is more robust.

Your `numpy_version_at_least('1.11') ` isn't very explicit though I think. You want it to say "1.11 or the lowest next available version" for your aarch64/ppcle migration. But I'd interpret `version_at_least` more like `pip` does (giving you the latest by default). 

I suspect that explicit is better than implicit here.

**Comment by hmaarrfk** [comment]
right, so if you want to be more rigid about the pinning, you would add a run_exports with a specific pinning

https://github.com/conda-forge/opencv-feedstock/pull/149/files#diff-e178b687b10a71a3348107ae3154e44cR177

Sorry for linking to a complicated recipe, but the idea is you would have

```
    build:
      run_exports:
        # don't pin the python version so hard.
        - {{ pin_subpackage('numpy', max_pin='x.x') }}
```
then you would control your ABI compatibility

**Comment by rgommers** [comment]
Okay, I see - your proposal would actually then give `numpy` more control over its own ABI, while letting packages that depend on it remove `- {{ pin_compatible('numpy') }}` from their `meta.yaml`. That does sound reasonable. I'll let more experienced conda-forge devs chime in now:)

**Comment by isuruf** [comment]
How would having `run_exports` solve the issue with the scikit-image repository?
